### PR TITLE
Add sensors for monitoring F-engine rx health

### PR DESCRIPTION
**Pull request by bmerry**
Checklist (if not applicable, edit to add `(N/A)` and mark as done):

- [x] If dependencies are added/removed: update ``setup.cfg`` and appropriate requirements files
- [x] (N/A) If modules are added/removed: use sphinx-apidoc to update files in ``doc/``
- [x] Ensure copyright notices are present and up-to-date
- [x] (N/A) If qualification tests are changed: attach a sample qualification report
- [x] (N/A) If design has changed: ensure documentation is up to date

Closes NGC-802, NGC-838 and NGC-839.

### CODE REVIEW

**Review by james-smith-za — CHANGES_REQUESTED** [review]
Overall I like it.
Just one nit-pick about a docstring which doesn't read so nicely (to me).

File: setup.cfg [hunk 1]
```diff
@@ -44,6 +44,7 @@ doc =
 
 test =
     katsdpsigproc[CUDA]
+    async-solipsism
```
[the remark this answers is not in the document]
Comment (bmerry) [reply setup.cfg:47]: Glad you like it :-) It tickled my fancy when I came up with it.

File: src/katgpucbf/utils.py [hunk 2]
```diff
@@ -111,6 +112,68 @@ def aggregate_compute(self) -> tuple[aiokatcp.Sensor.Status, DeviceStatus]:  # n
         return (aiokatcp.Sensor.Status.WARN, DeviceStatus.DEGRADED)
 
 
+class TimeoutSensorStatusObserver:
+    """Change the status of a sensor if it doesn't receive an update for a given time.
+
+    Do not directly attach or detach this observer from the sensor (it does
+    this internally). It can be constructed and then simply discarded if there
+    is no need to cancel it. Alternatively, keep a reference and use
```
[the remark this answers is not in the document]
Comment (bmerry) [reply src/katgpucbf/utils.py]: Is it better now?

**Comment by bmerry** [comment]
I'm thinking that a fixed timeout might not be the best approach, because with sufficiently large chunks or sufficiently low bandwidth, the time between chunks could be greater than the timeout. I think a better solution is to set the timeout to the expected time for, say, 10 chunks.

File: src/katgpucbf/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (james-smith-za) [reply src/katgpucbf/utils.py]: Much, thank you.

**Comment by james-smith-za** [comment]
> I'm thinking that a fixed timeout might not be the best approach, because with sufficiently large chunks or sufficiently low bandwidth, the time between chunks could be greater than the timeout. I think a better solution is to set the timeout to the expected time for, say, 10 chunks.

I don't object to this in principle. You'd just need a bit of logic to work out how long that ends up being.

File: src/katgpucbf/fgpu/recv.py [hunk 3]
```diff
@@ -210,9 +212,56 @@ def make_streams(
...
+                f"input{pol}-rx-unixtime",
+                "The timestamp (in UNIX time) of the last chunk of data received from the digitiser",
+                default=-1.0,
+                initial_status=aiokatcp.Sensor.Status.ERROR,
+            ),
+        ]
+        for sensor in timestamp_sensors:
+            TimeoutSensorStatusObserver(sensor, sensor_timeout, aiokatcp.Sensor.Status.ERROR)
+            sensors.add(sensor)
+        missing_sensors: list[aiokatcp.Sensor] = [
```
Comment (amishatishpatel) [inline src/katgpucbf/fgpu/recv.py:245]: I see you've added empty lines in fgpu's `Engine::_populate_sensors` to separate the concerns
* Might I request the same here after you've created the 'normal' `rx-time` sensors (please)?

**Review by amishatishpatel — APPROVED** [review]
Honestly, I can't find any holes in your plot here. Very cool, looking forward to using it.

Will give it another glance in the morning in case I missed anything/if James spots something.

File: test/fgpu/test_recv.py [hunk 4]
```diff
@@ -291,7 +293,24 @@ async def test_missing_heaps(
 class TestChunkSets:
     """Test :func:`.chunk_sets`."""
 
-    async def test(self, layout: Layout, caplog) -> None:  # noqa: D102
+    @pytest.fixture
+    async def sensors(self) -> aiokatcp.SensorSet:
+        """Receiver sensors."""
+        # This is an async fixture because make_sensors requires a running event loop
+        return make_sensors(1e6)  # Large timeout so that it doesn't affect the test
```
[the remark this answers is not in the document]
Comment (bmerry) [reply test/fgpu/test_recv.py:300]: > It also literally reads as if we're instructing it to make one million sensors.

I hope you're also reading that like [this](https://dailyhaymaker.com/wp-content/uploads/2014/01/drevil.png). Good point though, I'll update it.

**Review by james-smith-za — APPROVED** [review]
Only one observation but I don't feel strongly about it either way. I'm happy to merge.

File: src/katgpucbf/fgpu/recv.py [hunk 5]
```diff
@@ -210,9 +212,57 @@ def make_streams(
...
+                float,
+                f"input{pol}-rx-missing-unixtime",
+                "The timestamp (in UNIX time) when missing data was last detected",
+                default=-1.0,
+                initial_status=aiokatcp.Sensor.Status.NOMINAL,
+            )
+        ]
+        for sensor in missing_sensors:
+            TimeoutSensorStatusObserver(sensor, sensor_timeout, aiokatcp.Sensor.Status.NOMINAL)
+            sensors.add(sensor)
```
[the remark this answers is not in the document]
Comment (bmerry) [reply src/katgpucbf/fgpu/recv.py:257]: Someone might ask for rx-missing-timestamp.

**Pull request merged**